tensorflow.go

```diff
@@ -0,0 +1,172 @@
+	t := &Tensor{
+		tensor: tf.TF_NewTensor_wrapper(tf.TF_DataType(dataType), &(shape[0][0]), len(shape), data, int64(size)),
+	}
+
+	return t, nil
+}
+
+func encodeStrings(in []string) []byte {
+	size := 0
+	for _, s := range in {
+		size += 8
+		size += len(s)
+		size += len(proto.EncodeVarint(uint64(len(s))))
+	}
+
+	out := make([]byte, size)
+
+	dataPos := 8 * len(in)
+	data := out[dataPos:]
+	offset := 0
+	for i, s := range in {
+		inBytes := []byte(s)
+		binary.LittleEndian.PutUint64(out[i*8:], uint64(offset))
+		inLen := proto.EncodeVarint(uint64(len(s)))
+		offset += copy(data[offset:], inLen)
+		offset += copy(data[offset:], inBytes)
+	}
+	return out
+}
+
+func Constant(value interface{}) (*Tensor, error) {
+	switch v := value.(type) {
+	case string:
+		//		defer C.free(unsafe.Pointer(str))
+		buf := encodeStrings([]string{v})
+		t, err := newTensor(tf.DataType_DT_STRING, TensorShapeScalar, uintptr(unsafe.Pointer(&(buf[0]))), len(buf))
+		if err != nil {
+			return nil, err
+		}
+		t.buf = buf
+		return t, nil
+	default:
+		return nil, fmt.Errorf("tensorflow: unsupported type %T", value)
+	}
+}
+
+func statusToError(status tf.TF_Status) error {
+	code := tf.TF_GetCode(status)
+	message := tf.TF_Message(status)
+
+	if code != 0 {
+		return fmt.Errorf("tensorflow: %d: %v", code, message)
+	}
+	return nil
+}
+
+func (s *Session) Run(inputs map[string]*Tensor, outputs []string, targets []string) ([]*Tensor, error) {
+	inputNames := tf.NewStringVector()
+	inputValues := tf.NewTensorVector()
+	if inputs != nil {
+		for k, v := range inputs {
+			inputValues.Add(v.tensor)
+			inputNames.Add(k)
+		}
+	}
+	outputNames := tf.NewStringVector()
+	for _, n := range outputs {
+		outputNames.Add(n)
+	}
+
+	targetNames := tf.NewStringVector()
+	for _, n := range targets {
+		targetNames.Add(n)
+	}
+
+	outputValues := tf.NewTensorVector()
+	status := tf.TF_NewStatus()
+
+	tf.TF_Run_wrapper(s.session, inputNames, inputValues, outputNames, outputValues, targetNames, status)
+
+	result := []*Tensor{}
+	for i := int64(0); i < outputValues.Size(); i++ {
+		result = append(result, &Tensor{
+			tensor: outputValues.Get(int(i)),
+		})
+	}
+	return result, statusToError(status)
+}
+
+func (s *Session) ExtendGraph(graph *tf.GraphDef) error {
+	status := tf.TF_NewStatus()
+	buf, err := proto.Marshal(graph)
+	if err != nil {
+		return err
+	}
+	tf.TF_ExtendGraph(s.session, buf, status)
+	return statusToError(status)
+}
+
+func (t *Tensor) DataType() tf.DataType {
+	return tf.DataType(tf.TF_TensorType(t.tensor))
+}
+
+func (t *Tensor) NumDims() int {
+	return tf.TF_NumDims(t.tensor)
+}
+
+func (t *Tensor) Dim(n int) int {
+	return int(tf.TF_Dim(t.tensor, n))
+}
+
+func (t *Tensor) DataSize() int {
+	return int(tf.TF_TensorByteSize(t.tensor))
+}
+
+func (t *Tensor) Data() []byte {
+	length := t.DataSize()
+	return (*[1 << 30]byte)(unsafe.Pointer(tf.TF_TensorData(t.tensor)))[:length:length]
```

Comment:
Is 1 << 30 the right maximum size for a Tensor data?  I may be misreading, but the underlying c++ implementation has a 1<<40 max elements limit.  In any event, this should be commented - the 1<<30 is idiomatic to people who do cgo, but I'm not certain it's *correct*, and if it is, it should be explained for the next person to come along. :)